### DISCUSSION TITLE
Extension system is very confusing and limited

### DISCUSSION THREAD

**Discussion by SebastianStehle** [danny-avila/LibreChat#1891] [Feature Requests & Suggestions]
Hi,

Librechat has a lot of extension points but IMHO no proper plugin system.

How it typically works:

1. An plugin is something that can be installed, e.g. it runs at application start or installation time do inject additional functionality into the app. Plugins describe themselves using either just normal methods or they expose metadata through files. 
2. The application provides a set of extension points, usually through interfaces and some of kind of registries.
3. When an plugin is installed it has the option to register the functionality into the system and an extension can register multiple things.

There are several plugin mechanisms. You can support different AIs, tools and so on, but there is no unified way to register them.

I would like to implement a set of tools to talk with external systems, but even though they work together to achieve a task, you have to activate them one by one. They cannot share metadata, api keys and any configuration.

**Comment by danny-avila** [comment]
Plugins are not uniform in any way but I plan to make them more plug and play. The current implementations are more proof of concept and I’ll be delivering extensible frameworks as the features extend to all endpoints.
